### ISSUE TITLE
Information bypass from intermediate layers of DAG models

### ISSUE THREAD

**Issue by akde** [vlfeat/matconvnet#490]
Hello;
Thanks for this awesome toolbox.
I am extracting the features from segments of the images by using the pre-trained CNNs. I need 4096 features from segments of the image.  By using a code similar to this it is possible to do that:
(copied from http://www.vlfeat.org/matconvnet/quick/)

% load and preprocess an image
im = imread('peppers.png') ;
im_ = single(im) ; % note: 0-255 range
im_ = imresize(im_, net.meta.normalization.imageSize(1:2)) ;
im_ = im_ - net.meta.normalization.averageImage ;

% run the CNN
res = vl_simplenn(net, im_) ;

% show the classification result
scores = squeeze(gather(res(**end-2**).x)) ;

As you can see, I do not get the information from final result, instead I extract it from layer end-2 of the result.
My question is that is it possible to do the same thing in the DAG models? Since there is no vl_dagnn.m yet ( https://github.com/vlfeat/matconvnet/issues/88 ) I do not know what to do. Is it possible to do it? If yes can you please tell me how to do it.

Yours Sincerely...

**Comment by albanie** [comment]
Here's an example using `imagenet-vgg-f`, but a similar approach can be applied to other pre-trained models:

    % load a pretrained model
    net = load('imagenet-vgg-f');

    % convert it to a DagNN
    net = dagnn.DagNN.fromSimpleNN(net) ;

    % load and preprocess an image
    im = imread('peppers.png') ;
    im_ = single(im) ; % note: 0-255 range
    im_ = imresize(im_, net.meta.normalization.imageSize(1:2)) ; 
    im_ = im_ - net.meta.normalization.averageImage ;

    % find the layer containing the features of interest
    featureLayer = 'x19';
    
    % tell the network to preserve these variables
    net.vars(net.getVarIndex(featureLayer)).precious = true;

    % run the DagNN
    net.eval({'x0', im_}) 

    % retrieve the features you are interested in 
    features = net.vars(net.getVarIndex(featureLayer)).value;

    % remove singletons (dimensions 4096 x 1)
    features = squeeze(gather(features));

The names of the variables can differ from one model to another.  In this particular model, the name of the input variable is `x0` and the name of the features you described above is `x19`.   

Calling `net.print()` prints out an ascii diagram of all the layers in your network and the names of the variables connected to each layer - this is a convenient way of finding out which variable names you need.

For memory efficiency reasons the `DagNN` will clear the variables values in all intermediate layers once they are no longer needed during `net.eval()` by default.   Setting the `precious` value of the variables to `true` tells the network to preserve them so that they are not cleared at the end of the `eval()` function call.
